File: file_processing_ts/src/workflows.ts
import { proxyActivities, log } from '@temporalio/workflow';
import type { createDefaultActivities, createHostActivities } from './activities';

export const DEFAULT_TASK_QUEUE = 'file-processing-task-queue';

// Proxy for activities that run on any available worker.
const { download } = proxyActivities<ReturnType<typeof createDefaultActivities>>({
  taskQueue: DEFAULT_TASK_QUEUE,
  startToCloseTimeout: '5 minutes',
});

/**
 * Orchestrates the file-processing pipeline:
 *
 *  1. download  — runs on any worker; returns the local path and the
 *                 hostname of the worker that downloaded the file.
 *  2. process   — must run on the *same* host (file is on that disk).
 *  3. upload    — same host requirement.
 *
 * After download we know the host-specific task queue and create a second
 * proxy that targets it, pinning process and upload to the correct worker.
 */
export async function fileProcessingWorkflow(sourceUrl: string, destinationUrl: string): Promise<string> {
  // Step 1 — download on any available worker.
  const { hostTaskQueue, localPath } = await download(sourceUrl);
  log.info('File downloaded', { localPath, hostTaskQueue });

  // Steps 2 & 3 — target the exact host that holds the file.
  const { process, upload } = proxyActivities<ReturnType<typeof createHostActivities>>({
    taskQueue: hostTaskQueue,
    startToCloseTimeout: '5 minutes',
  });

  const processedPath = await process(localPath);
  log.info('File processed', { processedPath });

  await upload({ localPath: processedPath, destinationUrl });
  log.info('File uploaded', { destinationUrl });

  return 'OK';
}

File: file_processing/workflows.py
from datetime import timedelta

from temporalio import workflow

with workflow.unsafe.imports_passed_through():
    from activities import DEFAULT_TASK_QUEUE, DownloadResult, UploadInput, download, process, upload


@workflow.defn
class FileProcessingWorkflow:
    @workflow.run
    async def run(self, source_url: str, destination_url: str) -> str:
        # Step 1 — download on any available worker.
        result: DownloadResult = await workflow.execute_activity(
            download,
            source_url,
            start_to_close_timeout=timedelta(minutes=5),
            task_queue=DEFAULT_TASK_QUEUE,
        )
        workflow.logger.info(
            f"File downloaded to {result.local_path} on worker {result.host_task_queue}"
        )

        # Steps 2 & 3 — pin to the host that holds the file by targeting its
        # unique task queue.
        host_queue = result.host_task_queue

        processed_path = await workflow.execute_activity(
            process,
            result.local_path,
            start_to_close_timeout=timedelta(minutes=5),
            task_queue=host_queue,
        )
        workflow.logger.info(f"File processed: {processed_path}")

        await workflow.execute_activity(
            upload,
            UploadInput(local_path=processed_path, destination_url=destination_url),
            start_to_close_timeout=timedelta(minutes=5),
            task_queue=host_queue,
        )
        workflow.logger.info(f"File uploaded to {destination_url}")

        return "OK"
